File: apps/mobile/lib/src/api/password_security_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'authed_api.dart';

class PasswordRecoveryException implements Exception {
  const PasswordRecoveryException(this.statusCode, [this.responseBody]);
  final int statusCode;
  final String? responseBody;
}

class PasswordSecurityApi {
  PasswordSecurityApi({
    required this.baseUrl,
    AuthedApi? authedApi,
    http.Client? client,
  })  : _authedApi = authedApi ?? AuthedApi(baseUrl: baseUrl),
        _client = client ?? http.Client();

  final Uri baseUrl;
  final AuthedApi _authedApi;
  final http.Client _client;

  Future<void> requestPasswordReset(String email) {
    return _requestPasswordReset({'email': email.trim().toLowerCase()});
  }

  Future<void> requestPhonePasswordReset(String phone) {
    return _requestPasswordReset({'phone': phone.trim()});
  }

  Future<void> _requestPasswordReset(Map<String, String> contact) async {
    final response = await _client.post(
      baseUrl.resolve('/v1/auth/password/forgot'),
      headers: {'content-type': 'application/json'},
      body: jsonEncode(contact),
    );
    if (response.statusCode != 202) {
      throw PasswordRecoveryException(
        response.statusCode,
        response.body.isEmpty ? null : response.body,
      );
    }
  }

  Future<void> resetPassword(String token, String newPassword) async {
    final response = await _client.post(
      baseUrl.resolve('/v1/auth/password/reset'),
      headers: {'content-type': 'application/json'},
      body: jsonEncode({
        'token': token.trim(),
        'newPassword': newPassword,
      }),
    );
    if (response.statusCode != 200) {
      throw PasswordRecoveryException(
        response.statusCode,
        response.body.isEmpty ? null : response.body,
      );
    }
  }

  Future<List<SecuritySessionRecord>> listSessions(String accessToken) async {
    final payload = await _authedApi.get(
      '/v1/account/security/sessions',
      accessToken,
    );
    final values = payload['sessions'];
    if (values is! List) {
      throw const FormatException('Invalid security sessions response');
    }
    return values
        .map((value) => SecuritySessionRecord.fromJson(
              Map<String, dynamic>.from(value as Map),
            ))
        .toList(growable: false);
  }

  Future<int> changePassword(
    String accessToken,
    String currentPassword,
    String newPassword,
  ) async {
    final payload = await _authedApi.post(
      '/v1/account/security/password',
      accessToken,
      {
        'currentPassword': currentPassword,
        'newPassword': newPassword,
      },
    );
    return (payload['revokedSessions'] as num?)?.toInt() ?? 0;
  }

  Future<void> revokeSession(String accessToken, String sessionId) async {
    await _authedApi.post(
      '/v1/account/security/sessions/$sessionId/revoke',
      accessToken,
      const {},
    );
  }

  Future<int> revokeAllSessions(String accessToken) async {
    final payload = await _authedApi.post(
      '/v1/account/security/sessions/revoke-all',
      accessToken,
      const {},
    );
    return (payload['revokedSessions'] as num?)?.toInt() ?? 0;
  }
}

class SecuritySessionRecord {
  const SecuritySessionRecord({
    required this.id,
    required this.userAgent,
    required this.ipAddress,
    required this.createdAt,
    required this.expiresAt,
  });

  final String id;
  final String? userAgent;
  final String? ipAddress;
  final DateTime createdAt;
  final DateTime expiresAt;

  factory SecuritySessionRecord.fromJson(Map<String, dynamic> json) {
    return SecuritySessionRecord(
      id: json['id'] as String,
      userAgent: json['userAgent'] as String?,
      ipAddress: json['ipAddress'] as String?,
      createdAt: DateTime.parse(json['createdAt'].toString()).toUtc(),
      expiresAt: DateTime.parse(json['expiresAt'].toString()).toUtc(),
    );
  }
}

File: apps/web/lib/password-security-api.ts
import { getAuthed, postAuthed } from './authed-api';

const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL ?? 'http://localhost:4000';

export interface SecuritySessionRecord {
  id: string;
  userAgent: string | null;
  ipAddress: string | null;
  createdAt: string;
  expiresAt: string;
}

export class PasswordRecoveryError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly payload: Record<string, unknown>,
    readonly retryAfter?: number
  ) {
    super(message);
    this.name = 'PasswordRecoveryError';
  }
}

async function publicPost(
  path: string,
  body: Record<string, unknown>,
  challengeResponse?: string
): Promise<Record<string, unknown>> {
  const response = await fetch(`${apiBaseUrl}${path}`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      ...(challengeResponse ? { 'x-suqnaa-human-check': challengeResponse } : {})
    },
    body: JSON.stringify(body)
  });
  const payload = await response.json().catch(() => ({})) as Record<string, unknown>;
  if (!response.ok) {
    const parsedRetry = Number.parseInt(response.headers.get('retry-after') ?? '', 10);
    throw new PasswordRecoveryError(
      typeof payload.error === 'string' ? payload.error : 'Password recovery request failed',
      response.status,
      payload,
      Number.isFinite(parsedRetry) && parsedRetry > 0 ? parsedRetry : undefined
    );
  }
  return payload;
}

export async function requestPasswordReset(email: string, challengeResponse?: string): Promise<void> {
  await publicPost('/v1/auth/password/forgot', { email }, challengeResponse);
}

export async function resetPassword(token: string, newPassword: string): Promise<void> {
  await publicPost('/v1/auth/password/reset', { token, newPassword });
}

export async function listSecuritySessions(): Promise<SecuritySessionRecord[]> {
  const response = await getAuthed<{ sessions: SecuritySessionRecord[] }>('/v1/account/security/sessions');
  return response.sessions;
}

export async function changeAccountPassword(currentPassword: string, newPassword: string): Promise<number> {
  const response = await postAuthed<{ changed: true; revokedSessions: number }>(
    '/v1/account/security/password',
    { currentPassword, newPassword }
  );
  return response.revokedSessions;
}

export async function revokeSecuritySession(sessionId: string): Promise<void> {
  await postAuthed(`/v1/account/security/sessions/${sessionId}/revoke`, {});
}

export async function revokeAllSecuritySessions(): Promise<number> {
  const response = await postAuthed<{ revokedSessions: number }>(
    '/v1/account/security/sessions/revoke-all',
    {}
  );
  return response.revokedSessions;
}
